### PR TITLE
catalog: Remove storage usage from catalog

### PR DESCRIPTION
**Pull request by jkosh44**
`mz_storage_usage_by_shard` is currently a builtin table that, like most builtin tables, is backed by data in the durable catalog. Specifically, it is backed by the storage usage collection. The mz_storage_usage_by_shard table has the following columns:

  - `id uint8`
  - `shard_id text`
  - `size_bytes uint8`
  - `collection_timestamp timestamptz`

In order to have correct billing logic, we need the following guarantees:

  - `collection_timestamp` should be close to the real time that the row was written to the table.
  - collection_timestamp should be less than or equal to the logical timestamp that the row was written. i.e. If I run `SELECT * FROM mz_storage_usage_by_shard AS OF t` then I should see all rows that will ever exist s.t. `collection_timestamp <= t`.
  - collection_timestamp should be monotonically increasing.

The way we currently maintain these guarantees is by first generating the `collection_timestamp` via the timestamp oracle, and then generating the logical write timestamp via the timestamp oracle.

The storage usage collection takes up 99% of the space in the durable catalog in all Materialize deployments. It takes a significant amount of time to read this data into memory and iterate through it once it's in memory. This commit attempts to resolve this issue by removing the storage usage collection from the durable catalog. The source of truth for this data will move from the durable catalog to the `mz_storage_usage_by_shard`. This requires the following special handling of the `mz_storage_usage_by_shard` table:

  - It is not cleared on startup.
  - We do not allow any migrations to the table, because that would clear the contents.

The startup process is still blocked on reading in the entire `mz_storage_usage_by_shard` table and pruning old entries. A future commit will move this pruning into a background task.

Works towards resolving #28722

### Motivation
This PR adds a known-desirable feature.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/adapter/src/coord/message_handler.rs [hunk 1]
```diff
@@ -306,12 +308,25 @@ impl Coordinator {
...
-            ops.push(catalog::Op::UpdateStorageUsage {
-                shard_id: Some(shard_id.to_string()),
-                size_bytes: shard_usage.size_bytes(),
+            let id = match self.catalog().allocate_storage_usage_id().await {
+                Ok(id) => id,
+                Err(err) => {
+                    tracing::warn!("Failed to allocate ID for storage metrics: {:?}", err);
+                    return;
+                }
+            };
```
Comment (jkosh44) [inline src/adapter/src/coord/message_handler.rs:319]: If we crash here, then there can be gaps in the `id` column. I'm assuming that's not an issue.

File: src/adapter/src/coord/message_handler.rs [hunk 2]
```diff
@@ -306,12 +308,25 @@ impl Coordinator {
...
+        let mut ops = Vec::with_capacity(shards_usage.by_shard.len());
         for (shard_id, shard_usage) in shards_usage.by_shard {
-            ops.push(catalog::Op::UpdateStorageUsage {
-                shard_id: Some(shard_id.to_string()),
-                size_bytes: shard_usage.size_bytes(),
+            let id = match self.catalog().allocate_storage_usage_id().await {
+                Ok(id) => id,
+                Err(err) => {
+                    tracing::warn!("Failed to allocate ID for storage metrics: {:?}", err);
+                    return;
```
Comment (jkosh44) [inline src/adapter/src/coord/message_handler.rs:317]: This error path is surprising to me, but matches what we do if the catalog transaction fails.

**Comment by jkosh44** [comment]
I ran the following experiment on my staging account:

1. Create a staging env with 2,650,238 storage usage events pre-seeded in the catalog (the same amount as centerfield).
2. Measure the startup time of 3 restarts.
3. Apply this PR to the staging env.
4. Measure startup time of 3 restarts.

The average startup time went from 56.7645692227 seconds to 23.347768582 seconds.

I have another WIP PR that moves the storage usage pruning to the background (https://github.com/MaterializeInc/materialize/pull/29081). I ran the same experiment on that PR and the average startup time is 20.821180797 seconds.

File: src/adapter/src/coord.rs [hunk 3]
```diff
@@ -3522,6 +3540,55 @@ impl Coordinator {
...
+    /// This method will read the entire contents of [`MZ_STORAGE_USAGE_BY_SHARD`] into memory
+    /// which can be expensive.
+    ///
+    /// DO NOT call this method outside of startup. The safety of reading at the current oracle read
+    /// timestamp and then writing at whatever the current write timestamp is (instead of
+    /// `read_ts + 1`) relies on the fact that there are no outstanding writes during startup.
+    ///
+    /// In practice, this method is always safe to call because group commit has builtin fencing,
+    /// and we never commit retractions to [`MZ_STORAGE_USAGE_BY_SHARD`] outside of this method. So
+    /// we don't have to worry about double retractions. Still, better safe than sorry.
```
Comment (jkosh44) [inline src/adapter/src/coord.rs:3556]: @aljoscha this is my wall of text explaining why this is safe, please read it carefully and let me know if you agree.

File: src/adapter/src/coord.rs [hunk 4]
```diff
@@ -2942,6 +2952,14 @@ impl Coordinator {
             let idle_metric = self.metrics.queue_busy_seconds.with_label_values(&[]);
             let last_message_watchdog = Arc::clone(&last_message);
 
+            // TODO(jkosh44) Do this in the background.
+            if let Some(retention_period) = storage_usage_retention_period {
+                let prune_start = Instant::now();
+                info!("startup: coord serve: storage usage prune beginning");
+                self.prune_storage_usage_events_on_startup(retention_period).await;
+                info!("startup: coord serve: storage usage prune complete in {:?}", prune_start.elapsed());
```
Comment (jkosh44) [inline src/adapter/src/coord.rs:2960]: Annoyingly, this work happens on a completely different thread than all the other startup work, so it is not included in any of the other timings. Instead you have to add it manually to those timings.

File: src/adapter/src/coord.rs [hunk 5]
```diff
@@ -2347,10 +2349,17 @@ impl Coordinator {
...
+        let mz_storage_usage_by_shard_schema: SchemaSpecifier = self
+            .catalog()
+            .resolve_system_schema(MZ_STORAGE_USAGE_BY_SHARD.schema)
+            .into();
+        let is_storage_usage_by_shard = |entry: &CatalogEntry| -> bool {
+            entry.name.item == MZ_STORAGE_USAGE_BY_SHARD.name
+                && entry.name.qualifiers.schema_spec == mz_storage_usage_by_shard_schema
+        };
+        for system_table in entries.iter().filter(|entry| {
+            entry.is_table() && entry.id().is_system() && !is_storage_usage_by_shard(entry)
```
Comment (ParkMyCar) [inline src/adapter/src/coord.rs:2361]: Is it possible to panic/return an error if the fingerprint used to determine builtin migrations changes for the new `mz_storage_usage_by_shard` table?

File: src/adapter/src/coord.rs [hunk 6]
```diff
@@ -3522,6 +3540,55 @@ impl Coordinator {
...
+
+        let cutoff_ts = u128::from(read_ts).saturating_sub(retention_period.as_millis());
+        let mut expired = Vec::new();
+        for (row, diff) in current_contents {
+            assert_eq!(
+                diff, 1,
+                "consolidated contents should not contain retractions: ({row:#?}, {diff:#?})"
+            );
+            // This logic relies on the definition of `mz_storage_usage_by_shard` not changing.
+            let collection_timestamp = row.unpack()[3].unwrap_timestamptz();
```
Comment (ParkMyCar) [inline src/adapter/src/coord.rs:3578]: nit: maybe something like `row.unpack().get(3).expect("definition of mz_storage_by_shard changed")` to improve debugability?

File: src/catalog/tests/read-write.rs [hunk 7]
```diff
@@ -89,85 +84,6 @@ async fn test_confirm_leadership(
     Box::new(state2).expire().await;
 }
 
-#[mz_ore::test(tokio::test)]
-#[cfg_attr(miri, ignore)] //  unsupported operation: can't call foreign function `TLS_client_method` on OS `linux`
-async fn test_persist_get_and_prune_storage_usage() {
```
Comment (ParkMyCar) [inline src/catalog/tests/read-write.rs:94]: Making sure we retain and then prune storage usage updates is pretty important, is it possible to keep a test that exercises this?

**Review by ParkMyCar — APPROVED** [review]
In general looks good to me!

File: src/adapter/src/coord.rs (hunk 5 above)
Comment (jkosh44) [reply src/adapter/src/coord.rs:2329]: That was added here: https://github.com/MaterializeInc/materialize/blob/e1725417265f7295417bac728a51f75ff9c91f33/src/adapter/src/catalog/open.rs#L951

Unless I'm misunderstanding you.

File: src/adapter/src/coord.rs (hunk 6 above)
Comment (jkosh44) [reply src/adapter/src/coord.rs]: Done

File: src/adapter/src/coord.rs (hunk 5 above)
Comment (ParkMyCar) [reply src/adapter/src/coord.rs:2329]: Ahhh sorry I missed this, looks great!

File: src/catalog/tests/read-write.rs (hunk 7 above)
Comment (jkosh44) [reply src/catalog/tests/read-write.rs:94]: So we already have a bunch of end to end tests for storage usage, and I just added another one: https://github.com/MaterializeInc/materialize/blob/a7fac4d81e3827f497a1abc7e81672112e125b32/src/environmentd/tests/server.rs#L1107-L1588

Let me know if you think there's a scenario that's not already covered. I'll also loop in the testing team.

**Comment by jkosh44** [comment]
@MaterializeInc/testing Are there any additional tests that you think should be added? The PR description should be sufficient to fully understand the change, but let me know if anything is unclear. One potential test is to ensure that the `mz_internal.mz_storage_usage_by_shard` table is not empty after an upgrade, but I'm not exactly sure how we'd do that.

File: src/catalog/tests/read-write.rs (hunk 7 above)
Comment (jkosh44) [reply src/catalog/tests/read-write.rs:94]: Well looks like one of those tests are now failing, I'll take a look tomorrow.

**Comment by jkosh44** [comment]
@MaterializeInc/testing can you also take a look at the Nightly failures and let me know if anything is new: https://buildkite.com/materialize/nightly/builds/9213#_. They all look pretty unrelated to me.

File: src/adapter/src/coord.rs (hunk 3 above)
Comment (aljoscha) [reply src/adapter/src/coord.rs:3505]: Yes, my reasoning is:

Assumptions:
- group_commit confirms leadership, meaning it checks that we have not been fenced

1. by the time at which we call this method (prune), we have fenced out other processes from writing
2. we determine retractions
3. we try and apply them in group_commit

When 3. succeeds, no-one else can have tried sneaking in retractions.

File: src/catalog/src/durable/persist.rs [hunk 8]
```diff
@@ -1222,68 +1218,16 @@ impl OpenableDurableCatalogState for UnopenedPersistCatalogState {
     }
 }
 
-/// Certain large collections are only needed during startup. This struct helps us cache these
```
Comment (aljoscha) [inline src/catalog/src/durable/persist.rs:1225]: 🙌

File: src/storage-controller/src/lib.rs [hunk 9]
```diff
@@ -899,7 +899,7 @@ where
                     // Register the collection so our manager knows about it.
                     //
                     // NOTE: Maybe this shouldn't be in the collection manager,
-                    // and collection manager should only be responsble for
+                    // and collection manager should only be responsible for
```
Comment (aljoscha) [inline src/storage-controller/src/lib.rs:902]: Somehow you always find all these little typos... 👌

**Review by aljoscha — APPROVED** [review]
It's so nice to see this removed from the catalog!

**Comment by def-** [comment]
> @MaterializeInc/testing can you also take a look at the Nightly failures and let me know if anything is new: https://buildkite.com/materialize/nightly/builds/9213#_. They all look pretty unrelated to me.

Sorry for the late response: The failures are unrelated (and hopefully fixed on main)

**Review by def- — APPROVED** [review]

**Pull request merged**